exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp

```diff
@@ -0,0 +1,229 @@
+//
+// Copyright (c) 2018, University of Edinburgh
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//  * Redistributions of source code must retain the above copyright notice,
+//    this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of  nor the names of its contributors may be used to
+//    endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <exotica_ilqr_solver/ilqr_solver.h>
+
+REGISTER_MOTIONSOLVER_TYPE("ILQRSolver", exotica::ILQRSolver)
+
+namespace exotica
+{
+ILQRSolver::ILQRSolver() = default;
+ILQRSolver::~ILQRSolver() = default;
+
+void ILQRSolver::SpecifyProblem(PlanningProblemPtr pointer)
+{
+    if (pointer->type() != "exotica::DynamicTimeIndexedShootingProblem")
+    {
+        ThrowNamed("This ILQRSolver can't solve problem of type '" << pointer->type() << "'!");
+    }
+    MotionSolver::SpecifyProblem(pointer);
+    prob_ = std::static_pointer_cast<DynamicTimeIndexedShootingProblem>(pointer);
+
+    // initialize Gain matrices
+    Eigen::MatrixXd x = prob_->get_X(0), u = prob_->get_U(0);
+
+    int T = prob_->get_T();
+    K_gains_.assign(T, Eigen::MatrixXd(u.rows(), x.rows()));
```

Comment:
`u.rows()` should be the same as `prob_->get_num_controls()` (similar for x => num_positions + num_velocities), which I'd much prefer instead of getting u and x just for this sake. Also, does it make sense to assign zero gains rather than uninitialised?

Reply:
Zero gains => not really.. they are evaluated backwards one by one. This way it will complain if I've a mistake somewhere.